packageurl.go

```diff
@@ -184,39 +194,20 @@ func NewPackageURL(purlType, namespace, name, version string,
 // ToString returns the human-readable instance of the PackageURL structure.
 // This is the literal purl as defined by the spec.
 func (p *PackageURL) ToString() string {
-	// Start with the type and a colon
-	purl := fmt.Sprintf("pkg:%s/", p.Type)
-	// Add namespaces if provided
-	if p.Namespace != "" {
-		var ns []string
-		for _, item := range strings.Split(p.Namespace, "/") {
-			ns = append(ns, url.QueryEscape(item))
-		}
-		purl = purl + strings.Join(ns, "/") + "/"
-	}
-	// The name is always required and must be a percent-encoded string
-	// Use url.QueryEscape instead of PathEscape, as it handles @ signs
-	purl = purl + url.QueryEscape(p.Name)
-	// If a version is provided, add it after the at symbol
-	if p.Version != "" {
-		// A name must be a percent-encoded string
-		purl = purl + "@" + url.PathEscape(p.Version)
+	u := &url.URL{
+		Scheme:   "pkg",
+		RawQuery: p.Qualifiers.urlQuery(),
+		Fragment: p.Subpath,
 	}
 
-	// Iterate over qualifiers and make groups of key=value
-	var qualifiers []string
-	for _, q := range p.Qualifiers {
-		qualifiers = append(qualifiers, q.String())
-	}
-	// If there are one or more key=value pairs, append on the package url
-	if len(qualifiers) != 0 {
-		purl = purl + "?" + strings.Join(qualifiers, "&")
-	}
-	// Add a subpath if available
-	if p.Subpath != "" {
-		purl = purl + "#" + p.Subpath
-	}
-	return purl
+	// we use JoinPath and EscapePath as the behavior for "/" is only correct with that.
+	// We don't want to escape "/", but want to escape all other characters that are necessary.
+	u = u.JoinPath(p.Type, p.Namespace, strings.Join([]string{p.Name, p.Version}, "@"))
+	// write the actual path into the "Opaque" block, so that the generated string at the end is
+	// pkg:<path> and not pkg://<path>.
+	u.Opaque, u.Path = u.EscapedPath(), ""
```

Comment:
Why do we explicitly set u.Path to an empty string?

Reply:
Seems like it‘s technically not necessary, you‘re right:
https://go.dev/play/p/XqKNeqpDsrt

But i‘m not sure if the fact that u.Path is ignored if u.Opaque is set is „expected behavior“ or just an implementation detail.

The emptry string assignments makes it more obvious without knowing that :)